sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiFinalizeWritesDoFn.java

```diff
@@ -0,0 +1,147 @@
+  private final Counter batchCommitOperationsFailed =
+      Metrics.counter(StorageApiFinalizeWritesDoFn.class, "batchCommitOperationsFailed");
+
+  private Map<String, Collection<String>> commitStreams;
+  private final BigQueryServices bqServices;
+  @Nullable private DatasetService datasetService;
+
+  public StorageApiFinalizeWritesDoFn(BigQueryServices bqServices) {
+    this.bqServices = bqServices;
+    this.commitStreams = Maps.newHashMap();
+    this.datasetService = null;
+  }
+
+  private DatasetService getDatasetService(PipelineOptions pipelineOptions) throws IOException {
+    if (datasetService == null) {
+      datasetService = bqServices.getDatasetService(pipelineOptions.as(BigQueryOptions.class));
+    }
+    return datasetService;
+  }
+
+  @StartBundle
+  public void startBundle() throws IOException {
+    commitStreams = Maps.newHashMap();
+  }
+
+  @ProcessElement
+  @SuppressWarnings({"nullness"})
+  public void process(PipelineOptions pipelineOptions, @Element KV<String, String> element)
+      throws Exception {
+    String tableId = element.getKey();
+    String streamId = element.getValue();
+    DatasetService datasetService = getDatasetService(pipelineOptions);
+
+    RetryManager<FinalizeWriteStreamResponse, Context<FinalizeWriteStreamResponse>> retryManager =
+        new RetryManager<>(Duration.standardSeconds(1), Duration.standardMinutes(1), 3);
+    retryManager.addOperation(
+        c -> {
+          finalizeOperationsSent.inc();
+          return datasetService.finalizeWriteStream(streamId);
+        },
+        contexts -> {
+          LOG.error(
+              "Finalize of stream "
+                  + streamId
+                  + " failed with "
+                  + Iterables.getFirst(contexts, null).getError());
+          finalizeOperationsFailed.inc();
+          return RetryType.RETRY_ALL_OPERATIONS;
+        },
+        c -> {
+          LOG.info("Finalize of stream " + streamId + " finished with " + c.getResult());
+          finalizeOperationsSucceeded.inc();
+          commitStreams.computeIfAbsent(tableId, d -> Lists.newArrayList()).add(streamId);
+        },
+        new Context<>());
+    retryManager.run(true);
+  }
+
+  @FinishBundle
+  @SuppressWarnings({"nullness"})
+  public void finishBundle(PipelineOptions pipelineOptions) throws Exception {
+    DatasetService datasetService = getDatasetService(pipelineOptions);
+    for (Map.Entry<String, Collection<String>> entry : commitStreams.entrySet()) {
+      final String tableId = entry.getKey();
+      final Collection<String> streamNames = entry.getValue();
+      RetryManager<BatchCommitWriteStreamsResponse, Context<BatchCommitWriteStreamsResponse>>
+          retryManager =
+              new RetryManager<>(Duration.standardSeconds(1), Duration.standardMinutes(1), 3);
+      retryManager.addOperation(
+          c -> {
+            batchCommitOperationsSent.inc();
```

Comment:
FYI, BatchCommit will return detailed error info. Say if it thinks in the batch, some streams are already committed, then it will return the list of already committed streams here:
https://source.corp.google.com/piper///depot/google3/google/cloud/bigquery/storage/v1beta2/storage.proto;l=751?q=BatchCommitWriteStreams

To see if the operation really succeeded, commit_time in the response cannot be empty.

Reply:
Is there a reason for this? Sounds like an error-prone API, as most users will assume that a success response means success (usual ROT for RPCS - success means success. failure means it may have failed or may have succeeded).

Reply:
@yirutang Looking at the code, it seems a bit tricky. The current code structure handles failure/retry at the RPC layer. Unfortunately it appears the commit is not idempotent - retrying a commit will return STREAM_ALREADY_FINALIZED. This means that if one stream fails and we retry, we'll get stuck in an infinite retry loop. (we can't ignore STREAM_ALREADY_FINALIZED, since there may be other failed responses that need to be retried).

I think the simplest solution for now would be for me to stop batching commits and issue a separate commit per BatchCommit.

Reply:
Ok, retrying just the correct streams ended up not being difficult. PTAL